PAC suggest mode

kghatton opened the pull request:
Makes the integrations to Kafka, document-store and concordance API optional.
- Whenever any of these integrations is configured to be disabled, the healthcheck for that integration is omitted.
- Without Kafka, the mapper only supports mapping by posting native content to its HTTP `/map` endpoint.
- Without the document-store or concordance API, the mapper can only provide mapping in `suggest` mode, which is sufficient for transforming draft content in PAC for the purpose of suggesting annotations.

Commit:
:art: Don't require a ResilientClient, any Client will do.

Commit:
:hammer: Pass TransformationMode to FieldTransformer.

Commit:
:sparkles: Add SUGGEST mode, which omits certain transforms.

The transformations which call out to other services (document store and
concordances) and only manipulate tag values (leaving body text
unaltered) are relatively expensive and complex, and can be skipped.

Commit:
:sparkles: Accept mode as a query string parameter.

- Interpret the preview flag as PUBLISH or PREVIEW mode accordingly.
- In the event that preview and mode are both specified and conflict
with each other, mode wins as it's newer.
- Pass the mode on to the EomFileProcessor. Unify the processXXX methods
into a single method that calls the appropriate eligibility checks.
- Log entry into EomFileProcessor for traceability.

Commit:
:wrench: Make DSAPI, concordances and Kafka all optional in config.

By default these are configured as enabled (meaning that the features
are configured in the service and included in the health check), but
they can be disabled individually (in which case the feature(s) are not
configured in the service and are excluded from the health check). If
all features have been disabled, a dummy health check is configured to
prevent warning messages on startup.

Commit:
:art: :memo: Check for mode support; update README.

- If the document store or concordance API isn't enabled, then only
accept requests for suggest mode transformations, and gracefully reject
other modes.
- The UnsupportedTransformationModeException is reported to HTTP clients
with status 421 (Misdirected Request). Technically this is an HTTP/2
status code, but it seems the most appropriate (a request for preview or
publication should not have been made to a service thus configured).
- Updated the README to explain the transformation modes; also picked up
a few other things that were out of date.

Commit:
:whale: Configure UPP and PAC deployments.

- Inject enabling flags from environment vars in Dockerfile.
- Integrations to Kafka, document-store and concordance APIs are enabled
in UPP deployment, and disabled in PAC deployment.

peteclark-ft commented on src/main/java/com/ft/methodearticlemapper/resources/PostContentToTransformResource.java:
```diff
@@ -39,42 +48,58 @@ public PostContentToTransformResource(EomFileProcessor eomFileProcessor) {
+                    .error(e.getMessage())
+                    .exception(e);
+        } catch (UnsupportedTransformationModeException e) {
+            throw ClientError.status(SC_MISDIRECTED_REQUEST)
```
Is this valid to use 421 Misdirected here? From the specification, it sounds like 421 is mainly about http connections and disabling reuse for http/2?

kghatton replied:
Yeah, I'm not sure about it either. I was trying to avoid the ubiquitous `400` or `404`. This happens if you request a mapping with a mode that isn't supported in the current deployment, so, in that sense, the client has misdirected their request. I think I'd go for `404` as a Plan B - there's nothing syntactically incorrect about the request, so maybe "not found" is a reasonable response for this case.

peteclark-ft reviewed (approved):
I'm not 100% convinced about using 421 as the err status, but otherwise the code looks 🥇, and the tests run and pass for me

Commit:
:ok_hand: Respond with 404 for unsupported mapping mode.

Agreed that 421 is not intended for this case, so requesting a mode
that's unsupported is more like requesting a resource that's not found.

Commit:
Merge branch 'master' into feature/pac-suggest-mode

# Conflicts:
#	src/main/java/com/ft/methodearticlemapper/configuration/MethodeArticleMapperConfiguration.java

Commit:
:rocket: Deployment for PAC as well as UPP.

Commit:
:whale: Arbitrarily take one tag if there are many.

In general we don't expect multiple tags to point at the same commit,
but if, for example, we have to tag with different naming conventions to
deploy to different environments, then this may happen for legitimate
reason. Just take the first line of the git tag output, so that the
build works. In consequence the build-info may not contain the expected
tag, but it should contain an equivalent tag.

Commit:
:whale: All envs require API_HOST, others are specific to UPP or PAC.

Commit:
:whale: In fact DSAPI and concordance URLs must always have a value.

Commit:
:whale: Seems EVERYTHING has to be in the deployment.yaml :(

Commit:
:whale: env vars are strings.

Quite bored of this now.

Commit:
:lock: Add admin ingress.

Commit:
:whale: Addressing some FLU about Helm templating.

Commit:
:wrench: Fixing values syntax.

Commit:
:wrench: Use conditional templating for obtaining the Kafka proxy url.

peteclark-ft reviewed (approved):
I can see the ingress has registered itself in K8S dev, but since we have no ingress controllers running, I believe it'll have no effect.*

-----
*Source: https://kubernetes.io/docs/concepts/services-networking/ingress/
